### ISSUE TITLE
"$@" serves a purpose, no?

### ISSUE THREAD

**Issue by matthewpersico** [koalaman/shellcheck#1148]
#### For bugs
- Rule Id (if any, e.g. SC1000): SC2124
- My shellcheck version (`shellcheck --version` or 'online'): online
- [x] I tried on shellcheck.net and verified that this is still a problem on the latest commit
- [ ] It's not reproducible on shellcheck.net, but I think that's because it's an OS, configuration or encoding issue

#### For new checks and feature suggestions
- [x] shellcheck.net (i.e. the latest commit) currently gives no useful warnings about this
- [x] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

#### Here's a snippet or screenshot that shows the problem:

```sh
declare REPOS="$@"


```

#### Here's what shellcheck currently says:
```
Line 1	SC2148: Tips depend on target shell and yours is unknown. Add a shebang.
Line 1	SC2034: REPOS appears unused. Verify it or export it.
Line 1	SC2124: Assigning an array to a string! Assign as array, or use * instead of @ to concatenate.
```
#### Here's what I wanted or expected to see:

From https://www.gnu.org/software/bash/manual/html_node/Special-Parameters.html:
```
@
($@) Expands to the positional parameters, starting from one. When the expansion occurs within
 double quotes, each parameter expands to a separate word. That is, "$@" is equivalent to
 "$1" "$2" …. If the double-quoted expansion occurs within a word, the expansion of the first
 parameter is joined with the beginning part of the original word, and the expansion of the last
 parameter is joined with the last part of the original word. When there are no positional 
 parameters, "$@" and $@ expand to nothing (i.e., they are removed).
```

Since ```"$@"``` solves the problem of including spaces in your arguments, I don't think this statement should trigger a SC2124 warning.

**Comment by automorphism88** [comment]
Shellcheck's SC2124 warning is correct here. You're trying to assign an array to a regular variable. `$@` is an array, and `REPOS` is not an array. You should either do `declare -a REPOS=("$@")` in order to make `REPOS` an array, or use `declare REPOS="$*"` instead. `$*` is a single string, with each element of the array `$@` separated by the first character of `IFS` (by default, the space character). Unless there's a good reason why `REPOS` isn't an array, you probably want the former, since using an array will preserve the original word splitting, while `$*` will not, potentially creating problems if 

Protip: You can do the same thing with any array, by using `${foo[*]}` instead of `${foo[@]}`.

**Comment by matthewpersico** [comment]
Duh! Make REPOS and array. Of course. Sometimes you’re so used to doing things one way that you have blinders on. Later on I use $REPOS in a “for” loop:

for i in $REPOS

when I should be using

for i in ${REPOS[@]}

maybe with quotes (let’s see what shellcheck says).

Withdrawn.

**Comment by automorphism88** [comment]
> maybe with quotes (let’s see what shellcheck says)
It's generally a good idea to always double quote variables unless you know what you're doing and have a good reason not to. Without the double quotes, bash will perform word splitting on the variable (and glob expansion, where applicable), so the only reason not to use double quotes is if that's what you want to happen.

**Comment by matthewpersico** [comment]
Which is the reason I quoted $@ in the first place at the top of this thread. Duh, again.
